hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReader.java

```diff
@@ -165,6 +165,12 @@ public void run() {
         } catch (InterruptedException e) {
           LOG.trace("Interrupted while sleeping between WAL reads");
           Thread.currentThread().interrupt();
+        } catch (WALEntryFilterRetryableException e) {
```

Comment:
Instead of making WALEntryFilterRetryableException extend RuntimeException, could we extend IOException so that above `catch (IOException e)` block will handle the exception ?

Reply:
@shahrs87 That will make the `filter` method in `WALEntryFilter` to throw checked exception and it is not good to do that because of a breaking change with `throws` in the method signature.

Reply:
I see this exception is supposed to originate from [`filterEntry` method ](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReader.java#L355) and  `filter#filter` method doesn't throw any checked exception. Thats why we need to extend RuntimeException.
Can we add a `catch(WALEntryFilterRetryableException)` in `filterEntry` method and convert that into an IOException so that [`catch (IOException) here`](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReader.java#L158-L163) will handle the retry.

Reply:
IMO, converting it into `IOException` is not a good thing as well because `WALEntryFilterRetryableException` might have nothing to do with `IOException`, it can be because of any reason.

Reply:
Makes sense.  Can we club the handling of IOException and WALEntryFilterRetryableException 
 exceptions [here](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReader.java#L155-L163) something like `catch (IOException | WALEntryFilterRetryableException e)` since we want to take the same path as non eof exception takes.

Reply:
I'm not really happy with the inheritance from RTE here either but understand the compat issues. Is this the only place this exception will surface? Unhandled RTEs eventually go to the thread's unhandled exception handler, which is rarely what you want to happen.

Reply:
@apurtell Yes, this exception is specifically for WAL entry filter and this is the only place we filter the entry after reading from WAL.

Reply:
>  Can we club the handling of IOException and WALEntryFilterRetryableException

@sandeepvinayak  what do you think about this comment ?

Reply:
Oh I missed it, @shahrs87 addressed it now.